refactor: modernize recent `ByteType` usages and read/write functions

l0rinc opened the pull request:
Follow-up to https://github.com/bitcoin/bitcoin/pull/31524, applying a few of the review suggestions to the affected code, while enabling more trivial reads/writes to use the new `std::byte` interface.

l0rinc committed:
Modernize `ByteType` usage and simplify read/write functions

- Replaced `unsigned char` in `ByteType` concept with `uint8_t` for consistency.
- Used a single `ByteType auto*` parameter to reduce template boilerplate, as suggested by Russ.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
Use the new std::byte-enabled read/write methods

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

laanwj reviewed (approved):
Code review ACK 9908ab0581ec7a873514a09edb27a7cbaba3611d
Good to get rid of a few redundant `UCharCasts`.

l0rinc commented:
I haven't seen your commit, I've recreated it - do you want me to cherry-pick that here instead?

l0rinc commented:
If other reviewers think it's relevant enough we can simplify that draft PR of yours to focus on Spans.

theuni commented:
> Not sure about the first commit. There is no issue or risk with the existing code that is fixed by the style fixups, so to me it seems better to not change it, unless there is a compelling reason. Generally when it comes to style modernizations, my recommendation would be to include them in changes that touch the line anyway, or leave them as-is, or do them wholesale for the whole project (only if there is a good reason for them). But going out and doing them one-by one for specific subsections of the code, especially if the rationale for the change is "consistency" may not be a net benefit.

Agree with this. I don't have an issue with any of the individual changes in the first commit, but most of them would apply all over the codebase. And doing that all over would be a recipe for never-ending churn.

So unless the compiled output for the first commit looks any different, I'd rather not be doing random opinionated changes like this.

l0rinc commented:
> And doing that all over would be a recipe for never-ending churn

Isn't that what we're doing regularly though? Endlessly making the legacy codebase a tiny bit more readable, a tiny bit more modern, a tiny bit more maintainable and consistent?

> I'd rather not be doing random opinionated changes like this.

Previously, I've applied the following changes (as mentioned in the commit message), but I agree that a few of these could be reverted as being indeed opinionated:
- Replaced `unsigned char` in `ByteType` concept with `uint8_t` for consistency.
- Used a single `ByteType auto*` parameter to reduce template boilerplate, as suggested by Russ.
- ~Replaced hardcoded `memcpy` lengths (2,4,8) with `sizeof(x)` and switched to `std::memcpy`.~
- ~Added `noexcept` to functions.~
- Used brace initialization instead of `=` for consistency.
- Marked values in write methods as `const` to make explicit what is being written.

Done in https://github.com/bitcoin/bitcoin/compare/9908ab0581ec7a873514a09edb27a7cbaba3611d..bf5baae88e350365cc07d1fbc38b47ce7f701e13

theuni commented:
> > And doing that all over would be a recipe for never-ending churn
> 
> Isn't that what we're doing regularly though? Endlessly making the legacy codebase a tiny bit more readable, a tiny bit more modern, a tiny bit more maintainable and consistent?

Not change for change's sake, no. Please consider this from the POV of reviewers and maintainers. A constant stream of changes like that is exhausting. It would distract from larger improvements.

I think @maflcko's comment above is a good description of the culture of this project: modernize the code when you touch it, and if it's important enough to apply to the rest of the codebase, do it in one shot and be done with it.

We (and bots) could fill endless PRs with things that make things a tiny bit more modern/maintainable/consistent, but we don't for the sake of the sanity of our peers.

l0rinc commented:
This PR was a follow-up from the comments on Marco's change.
The main changes are Russ's template simplification and my suggestion to const the writes to obviate which we're modifying.

l0rinc commented:
> Though for objects, it is mostly style

It's not purely a stylistic change - it clarifies which parameter remains unchanged (i.e., it isn’t a "return value").

It seems there’s a strong preference against this change, so I don't mind adjusting the approach or closing the PR altogether.

l0rinc commented:
I'm not sure these arguments would hold in other cases (e.g. no compiler safety needed since the tests would fail otherwise), but I've removed the consts as well, basically only leaving @ryanofsky's changes.

hodlinator commented:
> (For reference, the commit that drops the unused casts ([f23d6a5](https://github.com/bitcoin/bitcoin/commit/f23d6a572d62ff640a28718ff0771c3f0f87f365)) has already been merged.)

Actual commit: https://github.com/bitcoin/bitcoin/commit/fad4032b219e58300f8d0a74bbcf38ef26fa89d0
